src/icp_canister_backend/src/lib.rs

```diff
@@ -128,4 +128,54 @@ pub async fn deposit(user: Principal, timelock: u64) -> Result<(), types::Deposi
     Ok(())
 }
 
+#[ic_cdk::update]
+pub async fn withdraw(deposit_id: u64) -> Result<Nat, types::WithdrawError> {
+    let caller = ic_cdk::api::caller();
+    let now = ic_cdk::api::time();
+
+    let deposit = DEPOSITS.with(|deposits| deposits.borrow().get(&deposit_id).map(|d| d.clone()));
+    let deposit = match deposit {
+        Some(d) => d,
+        None => return Err(types::WithdrawError::NoDeposit),
+    };
+
+    if deposit.principal != caller {
+        return Err(types::WithdrawError::NotOwner);
+    }
+
+    if now < deposit.unlocktime {
+        return Err(types::WithdrawError::TimelockNotExpired);
+    }
+
+    let ledger_principal = TOKEN_ID.with(|cell| cell.borrow().get().clone());
+    let transfer_amount = if deposit.amount.0 > TRANSFER_FEE.0 {
```

Comment:
In the deposit logic, we check if deposit amount > TRANSFER_FEE, but it may be less then 2 * TRANSFER_FEE, so this deposit, couldn't be withdrawed ever.
So, we have to modify deposit logic and check if deposit is more then MINIMUM_DEPOSIT_AMOUNT which should be 0,00001 of token.
Then we can remove this check here, because we are sure that all deposits are big enough.

Also, add a test for the deposits - to check that it is not possible to deposit if amount is smaller then MINIMUM_DEPOSIT_AMOUNT.